proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,414 @@
+
+* There is no way to avoid punning if the user imports the code that makes the
+  use of it.
+  Consider this import statement: `import Data.Proxy`. It is impossible to
+  distinguish between `Proxy` the type and `Proxy` the data constructor.
+
+#### Beginner confusion
+
+Even without Dependent Haskell an argument can be made for pun-free code:
+punning can be a source of confusion for beginners. The difference between
+the terms namespace and the types namespace can be hard to understand at first,
+especially when things like `()` or `[a]` are used ([as seen in this StackOverflow question](https://stackoverflow.com/questions/16892570/what-is-in-haskell-exactly)).
+
+### Solution Overview
+
+#### Namespace qualified imports
+
+To help programmers deal with the external code that uses punning we propose
+to introduce namespace qualified import syntax, guarded behind
+`-XExplicitNamespaces` extension. The syntax introduces two markers for
+qualified imports, `data` and `type`:
+
+```haskell
+import Data.Proxy type qualified as T   -- import only the type namespace
+import Data.Proxy data qualified as D   -- import only the data namespace
+```
+
+#### Punning usage warnings
+
+We also propose to introduce two new warnings to GHC: `-Wpuns` and
+`-Wpun-bindings`.
+
+* `-Wpuns` warns the user about the usage of punning at use sites.
+
+* `-Wpun-bindings` warns the user about the introduction of punning at binding
+  sites.
+
+These warnings aim to help the user to avoid using punning in their codebase.
+
+To determine whether some binding or use site takes advantage of punning we
+ask the question: "If Haskell had a single unified namespace, would that change
+the lookup result?". If the answer is yes, then the code uses punning:
+
+```haskell
+data T = T -- with a single unified namespace, T the data would have a name clash with T the type.
+
+a = 15
+
+f :: forall a. a -> a -- no warning, because with a single unified namespace 'forall a.' would shadow the top level 'a', so 'a' in 'a -> a' would still refer to the forall bound 'a'.
+```
+
+There are more detailed examples in Examples section.
+
+## Proposed Change Specification
+
+* To disambiguate the namespaces of identifiers from modules that use punning,
+  the `-XExplicitNamespaces` extension is extended with new syntax:
+
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+
+* Introduce a new warning, `-Wpun-bindings` and add it to `-Weverything`.
+  The warning is triggered by any name binding that would clash with another
```

Comment:
Could you add more detail to "clash"? For example, I *think* this warning wouldn't trigger if the two definitions (one in the type namespace and one in the term namespace) are in different modules. I think you mean that the warning is triggered whenever a definition would be *rejected* if there were only one namespace. So definitions in different modules don't trigger it. What about shadowing circumstances? For example:

```hs
id @a (x :: a) = (\a -> a) x
```

Would that trigger the warning? Does it depend on whether we're warning for shadowed identifiers?

Reply:
>  I think you mean that the warning is triggered whenever a definition would be rejected if there were only one namespace

Yes, that is the intention. If a definition is not rejected, it wouldn't trigger the warning.
That's why `-Wpuns-bindings` example 4 triggers no warnings.

Shadowing is fine and does not trigger this warning, I think your example is similar to the last code snippet of `-Wpun-bindings` example 1, unless I missed something:

```haskell
f :: t -> ()
f @a = \a -> ()
```

Reply:
Looks great now. Thanks.